docs/developer/architecture/logs.rst

```diff
@@ -197,6 +197,527 @@ overview page.
+
+.. code-block:: yaml
+
+  serviceMonitor:
+    enabled: true
+    additionalLabels:
+      release: "prometheus-operator"
+
+For Fluent Bit, we will need to define a ``ServiceMonitor`` object:
+
+.. code-block:: yaml
+
+  apiVersion: monitoring.coreos.com/v1
+  kind: ServiceMonitor
+  metadata:
+    labels:
+      app: fluent-bit
+      release: prometheus-operator
+    name: fluent-bit
+    namespace: metalk8s-logging
+  spec:
+    endpoints:
+    - path: /api/v1/metrics/prometheus
+      port: http-metrics
+    namespaceSelector:
+      matchNames:
+      - metalk8s-logging
+    selector:
+      matchLabels:
+        app: fluent-bit
+
+We also need to define alert rules based on metrics exposed by these services.
+This is done deploying new ``PrometheusRules`` object in ``metalk8s-logging``
+Namespace:
+
+.. code-block:: yaml
+
+  apiVersion: monitoring.coreos.com/v1
+  kind: PrometheusRule
+  metadata:
+    labels:
+      release: prometheus-operator
+    name: loki.rules
+    namespace: metalk8s-logging
+  spec:
+    groups:
+    - name: loki.rules
+      rules: <RULES DEFINITION>
+
+There is some recording and alert rules defined in the `Loki repository`_ that
+could be used as a base, then we could enrich these rules later when we will
+have better operational knowledge.
+
+.. _Loki repository: https://github.com/grafana/loki/tree/master/production/loki-mixin
+
+Grafana
+"""""""
+
+To be able to query logs from Loki, we need to add a Grafana datasource,
+this is done adding a ConfigMap ``loki-grafana-datasource`` in
+``metalk8s-monitoring`` Namespace as follows:
+
+.. code-block:: yaml
+
+  apiVersion: v1
+  kind: ConfigMap
+  metadata:
+    name: loki-grafana-datasource
+    namespace: metalk8s-monitoring
+    labels:
+      grafana_datasource: "1"
+  data:
+    loki-datasource.yaml: |-
+      apiVersion: 1
+      datasources:
+      - name: Loki
+        type: loki
+        access: proxy
+        url: http://loki.loki.svc.cluster.local:3100
+        version: 1
+
+To display the logs we also need a dashboard, adding a ConfigMap
+``loki-logs-dashboard`` in Namespace ``metalk8s-monitoring``:
+
+.. code-block:: yaml
+
+  apiVersion: v1
+  kind: ConfigMap
+  metadata:
+    name: loki-logs-dashboard
+    namespace: metalk8s-monitoring
+    labels:
+      grafana_dashboard: "1"
+  data:
+    loki-logs.json: <DASHBOARD DEFINITION>
+
+For the dashboard we will use a view with a simple log panel and variables
+representing labels to filter on.
+Since journal and kubernetes logs will not have the same labels, we could
+either have 2 distinct dashboards or 2 log panels in the same.
+An example of what we want is `Loki dashboard`_.
+
+.. _Loki dashboard: https://github.com/grafana/loki/blob/master/production/loki-mixin/dashboard-loki-logs.json
+
+.. note::
+
+  The ``grafana_datasource: "1"`` and ``grafana_dashboard: "1"`` labels are
+  what is used by the Prometheus Operator to retrieve datasource and dashboard
+  for Grafana.
+
+Loki Volume Purge
```

Comment:
One really strong message we received from field experts (cc @xaf-scality @bkettler) is that even though the volume is filling up at a dangerous rate, it should never get to the point where it's full.

Monitoring and alerting are paramount for an admin to make sure the system is stable, but a broken component or a DDoS attack over the week-end could result in ingestion increase, which shouldn't break the log system to a point where complex manual operations are required.

Depending on the cause, the logging pipeline could (should?) first pause automatically, and then:

- storage should be adjusted (as you said, by lowering retention or increasing capacity)
- outliers should be detected and silenced / removed from the pipeline (manually or automatically, depends on what we can do)
- ...other ideas?

Reply:
First, for the pause in the log ingestion, I think it's already made this way in fluent-bit.
If an output plugin is unable to send data to a backend, it'll keep the data in memory and if the buffer is full, then it tells the input plugin to pause, see https://fluentbit.io/documentation/0.13/configuration/backpressure.html for more details.

For the storage resizing, we can't do much except providing some alerting (with prediction and also real time monitoring), documentation (through runbook for example) and good sizing rules.

For the outliers... I'm not sure how to achieve this automatically, it means we need to be able to analyze log flow and detect patterns... I'm not sure that's what we want in our simple logging stack.
To me, the pause + alert + documentation should be enough.
Maybe we can also provide some documentation to explain how to do basic filtering to be sure unwanted logs are dropped at collector level.

Otherwise, we could add an extra option in Loki configuration that would allow us to do purging based on size, which could be done with 2 options (to ensure we will not remove all logs if there is a crazy amount of logs):
- % of space used: we purge oldest logs if space is greater than a certain threshold 
- minimum days of retention: we purge logs older than minimum days of retention
It means, for example, you can have a retention of 2 weeks of logs, with a minimum of 3 days and if it's still not sufficient, fluent-bit will pause the flow.
I just think that if we go with automatic purge based on size, we need to be careful and trigger alerts each time we purge, otherwise operators will think they have 2 weeks of logs and will never be aware that there is a sizing issue on their platform.